### PR TITLE
[755] zio-schema breaking update

### PR DESCRIPTION
**Pull request by m-kalai**
https://github.com/zio/zio-redis/issues/755

* getting rid of implicit BinaryCodec where it's not needed
* introducing CodecSupplier as a way to provide BinaryCodec implementation used for arbitrary input/output

### CODE REVIEW

**Review by ioleo — APPROVED** [review]
:+1: LGTM

File: redis/src/test/scala/zio/redis/codecs/ProtobufCodecSupplier.scala [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package zio.redis.codecs
+
+import zio.redis.CodecSupplier
+import zio.schema.Schema
+import zio.schema.codec.{BinaryCodec, ProtobufCodec}
+
+object ProtobufCodecSupplier extends CodecSupplier {
+  implicit def codec[A: Schema]: BinaryCodec[A] = ProtobufCodec.protobufCodec
+}
```
Comment (mijicd) [inline redis/src/test/scala/zio/redis/codecs/ProtobufCodecSupplier.scala:9]: Perhaps it's worth exposing it via codec supplier's companion constructor, e.g.

```scala
object CodecSupplier {
  def protobuf: CodecSupplier = new CodecSupplier { ... }
}
```

The same applies for string codec.

File: redis/src/main/scala/zio/redis/RedisEnvironment.scala [hunk 2]
```diff
@@ -16,9 +16,12 @@
...
 
+import zio.schema.Schema
 import zio.schema.codec.BinaryCodec
 
 private[redis] trait RedisEnvironment {
-  protected def codec: BinaryCodec
+  protected def codec: CodecSupplier
   protected def executor: RedisExecutor
+
+  implicit def summonCodec[A: Schema]: BinaryCodec[A] = codec.codec
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/RedisEnvironment.scala:26]: Is it necessary?

File: redis/src/main/scala/zio/redis/CodecSupplier.scala [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+package zio.redis
+
+import zio.redis.codecs.StringUtf8Codec
+import zio.schema.Schema
+import zio.schema.codec.BinaryCodec
+
+trait CodecSupplier {
+  implicit def codec[A: Schema]: BinaryCodec[A]
+}
```
[the remark this answers is not in the document]
Comment (m-kalai) [reply redis/src/main/scala/zio/redis/CodecSupplier.scala:25]: 1. it's already called `CodecSupplier` or am i not seeing the difference?
2. ok sure
3. let me check if it's not used somewhere, if not i'll remove

File: redis/src/main/scala/zio/redis/RedisEnvironment.scala (hunk 2 above)
Comment (m-kalai) [reply redis/src/main/scala/zio/redis/RedisEnvironment.scala:26]: yeah, we have to "summon" the implicit codec, without this you get like 600 error when trying to compile, is there a better way you think?

File: redis/src/main/scala/zio/redis/CodecSupplier.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mijicd) [reply redis/src/main/scala/zio/redis/CodecSupplier.scala]: Sorry, typo :man_facepalming: I wanted to write `CodecProvider`.

**Review by mijicd — APPROVED** [review]

**Pull request merged**